plugins/community/nipunn1313-mypy-grpc/v3.5.0/buf.plugin.yaml

```diff
@@ -7,3 +7,22 @@ spdx_license_id: Apache-2.0
 license_url: https://github.com/nipunn1313/mypy-protobuf/blob/v3.5.0/LICENSE
 output_languages:
   - python
+registry:
+  python:
+    package_type: "stub-only"
+    # https://github.com/nipunn1313/mypy-protobuf/blob/v3.5.0/setup.cfg#L20
+    requires_python: ">=3.8"
+    deps:
+      # https://github.com/nipunn1313/mypy-protobuf/tree/v3.5.0#requirements-to-run-typecheckers-on-stubs-generated-by-mypy-protobuf
+      # > Earlier releases might work, but aren't tested.
+      # TODO: Should we just remove the version specifiers, then?
+      - "types-protobuf >= 4.23.0.2"
+      - "protobuf >= 4.23.4"
+      # https://github.com/nipunn1313/mypy-protobuf/tree/v3.5.0#to-run-typecheckers-on-code-generated-with-grpc-plugin---youll-additionally-need
+      # > Earlier releases might work, but aren't tested
+      # TODO: Should we just remove the version specifiers, then?
+      - "grpcio >= 1.56.2"
+      # https://pypi.org/project/grpcio-tools/
+      - "grpcio-tools >= 1.56.2"
+      # https://pypi.org/project/grpc-stubs/
+      - "grpc-stubs >= 1.53.0.2"
```

Comment:
Over time what is the best way to keep these up to date with newer dependencies? Is it something we can have the fetcher do as it upgrades things?

Reply:
Unfortunately, there's not a great way that I'm aware of - there's not a way that these are consistently encoded across Python packages. In this _particular_ package's case, we might be able to read them from the [`test_requirements.txt`](https://github.com/nipunn1313/mypy-protobuf/blob/main/test_requirements.txt#L6-L8) file, but other plugins (like the protocolbuffer/{python,pyi} ones) don't have any encoded information about dependencies to use the generated code.

One other wrinkle with Python dependencies is that libraries are typically expected to keep their dependencies unpinned - that is, not use `==`. They can set a minimum version (like what we're doing here), but this _does_ mean that we're now forcing users using this package to install a version greater than or equal to what we're specifying, which isn't necessarily a strict limit.

That's one of the reasons why I have the `TODO` above here - it's not clear to me if we should even include the `>=`, or leave it to consumers to specify the version of these deps that they want?

Reply:
I've been thinking through this off-and-on this week and really haven't come up with much better. My preference would be that we start with:

* use the ranges specified by the plugin author, if available
  * we can relax these if people complain and want to figure out their own versions, see below
* otherwise, leave the version range open and leave it to the user to install the version they need.
  * we can tighten these up if we realize that there's a plugin that _always_ requires something from a certain version of a dependency

Obviously, we can/should tighten these up if we get more information, but ISTM that this is just sorta the convention in the ecosystem: libraries should declare pretty loose requirements so that upstream applications can figure out the versions they need. (See https://hynek.me/articles/semver-will-not-save-you/ section "Unintended Consequence: Version Conflicts" for a bit of related reading, but the core thing I'm gesturing at is leaving it up to the user to figure these things out.)